contracts/external/cw-payroll/src/balance.rs

```diff
@@ -0,0 +1,185 @@
+use cosmwasm_std::{Coin, StdError,OverflowOperation::Sub,OverflowError};
+use cw20::{Balance, Cw20CoinVerified};
+use cosmwasm_schema::cw_serde;
+
+use crate::error::GenericError;
+
+#[cw_serde]
+#[derive(Default)]
+pub struct GenericBalance {
+    pub native: Vec<Coin>,
+    pub cw20: Vec<Cw20CoinVerified>,
+}
```

Comment:
1. vectors aren't a great data structure for indexing based on keys.
2. every time we load this from storage, we pay gas for all of the data in it. pretty trivial to DOS the contract, this being the case. check out "attempt two" in this essay for an example of that: https://github.com/DA0-DA0/dao-contracts/discussions/523

how about we move to using cosmwasm maps for these? this might be a nice example to work from: https://github.com/public-awesome/ics721/blob/main/packages/cw-pause-once/src/lib.rs#L23-L26

Reply:
> 1. vectors aren't a great data structure for indexing based on keys.
> 2. every time we load this from storage, we pay gas for all of the data in it. pretty trivial to DOS the contract, this being the case. check out "attempt two" in this essay for an example of that: [The three horsemen of NFT staking and voting power #523](https://github.com/DA0-DA0/dao-contracts/discussions/523)
> 
> how about we move to using cosmwasm maps for these? this might be a nice example to work from: https://github.com/public-awesome/ics721/blob/main/packages/cw-pause-once/src/lib.rs#L23-L26

yeah, agree with you, just we considering not holding much data in the vectors, but still you are right, we need to modify this ones)

Reply:
> that seem to create their own map?

that may have been a confusing explanation :) they don't actually make a `Map`. just trying to suggest that we refactor this to look like that type. for example:

```rust
pub struct GenericBalance {
    pub native: Map<String, Uint128>,
    pub cw20: Vec<String, Uint128>,
}
```

Reply:
After discussion - Let's roll with single token amounts here.
As @0xekez pointed out, a lot of fun edge cases open up with multiple tokens.
Since you can create multiple "streams", we can then force 1 stream to be holding (1 native asset, 1 cw20 asset). 
If there are real use cases to go beyond this, it can be a future upgrade with adequate research & test cov